File: src/builder/transpilers/latex/04-center.ts
import { blockContent, hasClass } from '../_ast-utils.js';

/**
 * Transpiler AST: transforma Divs con clase .center al entorno
 * \begin{center}...\end{center} en LaTeX.
 *
 * Emite RawBlocks de apertura/cierre alrededor de los bloques internos
 * nativos: pandoc los convierte en la misma pasada, con cero procesos extra.
 */

export const type = 'ast' as const;

function processCenter(block: Record<string, unknown>): unknown[] {
  const content = blockContent(block);
  return [{ t: 'RawBlock', c: ['latex', '\\begin{center}'] }, ...content, { t: 'RawBlock', c: ['latex', '\\end{center}'] }];
}

export async function transform(ast: Record<string, unknown>): Promise<Record<string, unknown>> {
  const blocks = ast.blocks as unknown[];

  const newBlocks: unknown[] = [];
  for (const block of blocks) {
    if (
      typeof block === 'object' &&
      block !== null &&
      (block as Record<string, unknown>).t === 'Div' &&
      hasClass(block as Record<string, unknown>, 'center')
    ) {
      newBlocks.push(...processCenter(block as Record<string, unknown>));
    } else {
      newBlocks.push(block);
    }
  }
  ast.blocks = newBlocks;
  return ast;
}

File: src/lib/resources/transpilers/latex/04-center.lua
-- Convierte Div.center al entorno \begin{center}...\end{center} (formato LaTeX).
-- Reemplaza al transpiler TS latex/04-center (Fase 6).
-- Uso: pandoc --from json --to latex --lua-filter latex/04-center.lua

local function has_class(block, cls)
  if block.t ~= 'Div' then return false end
  for _, c in ipairs(block.classes) do
    if c == cls then return true end
  end
  return false
end

function Pandoc(doc)
  local blocks = {}
  for _, block in ipairs(doc.blocks) do
    if has_class(block, 'center') then
      table.insert(blocks, pandoc.RawBlock('latex', '\\begin{center}'))
      for _, b in ipairs(block.content) do
        table.insert(blocks, b)
      end
      table.insert(blocks, pandoc.RawBlock('latex', '\\end{center}'))
    else
      table.insert(blocks, block)
    end
  end
  doc.blocks = blocks
  return doc
end
